module/zfs/metaslab.c

```diff
@@ -2102,6 +2283,79 @@ metaslab_condense(metaslab_t *msp, uint64_t txg, dmu_tx_t *tx)
 	msp->ms_condensing = B_FALSE;
 }
 
+/*
+ * Synchronize the allocation-by-category data into our class instance.
+ * Reset counters for the next txg sync pass.
+ * Class instance data will get synced after all top-level vdevs are synced
+ */
+void
+metaslab_sync_alloc_stats(metaslab_t *msp, uint64_t txg)
+{
+	metaslab_class_t *mc = msp->ms_group->mg_class;
+	uint64_t index = txg & TXG_MASK;
+
+	if (mc->mc_alloc_stats_obj == 0 ||
+	    mc->mc_alloc_stats_data.cas_birth_txg == 0)
+		return;
+
+	VERIFY(txg <= spa_final_dirty_txg(mc->mc_spa));
+
+	ms_alloc_stats_t *mas = &msp->ms_alloc_stats[index];
+
+	mutex_enter(&mc->mc_lock);
+	class_alloc_stats_phys_t *cas = &mc->mc_alloc_stats_data;
+
+	/* dedup data allocated */
+	cas->cas_dedup_data.as_bytes += mas->mas_dedup_bytes;
+	cas->cas_dedup_data.as_blocks += mas->mas_dedup_blocks;
+	ASSERT((int64_t)cas->cas_dedup_data.as_bytes >= 0);
+	ASSERT((int64_t)cas->cas_dedup_data.as_blocks >= 0);
```

Comment:
you could use `ASSERT3S` to do the cast for you (and get additional info on failure)

Reply:
updated to `ASSERT3S`